### DISCUSSION TITLE
Global (mouse) event listeners and drag preview in Avalonia possible?

### DISCUSSION THREAD

**Discussion by BreyerW** [AvaloniaUI/Avalonia#12609] [Q&A] [answered]
By global i truly mean global eg when im moving/dragging or realeasing/dropping something with mouse BEYOND window app.

Currently, I can listen to mouse move events globally only with `PositionChanged` of `Window` when dragging it but thats too little for my needs since `Pointer*` events gets eaten and i have particular need for `PointerRelease` but wouldnt said no to `PointerMoved` as well. I tried assigning handler to `TopLevel`/`Window`, doesnt work, overriding `OnPointer*` events on custom `Controls`, doesnt work either. Same even when i capture control with `e.Pointer.Capture(control)`

I have found [https://sharphook.tolik.io/v5.0.0/articles/hooks.html](https://sharphook.tolik.io/v5.0.0/articles/hooks.html) which seems to do what i need so answer to this isnt crucial as i can workaround it but i would love to solve my problems purely with Avalonia (since it has its own input managers)

If it isnt possible right now is it planned for future releases? It could be hidden behind Desktop lifetime class as i imagine it wouldnt work on most mobiles and i need only desktop support for my needs.

As for second question - its partially related to first cause currently my Dock Manager is based on dragging floating, undecorated window when undocking from TabControl but im seriously considering instead rendering `TabItem` to bitmap and use that bitmap as preview when dragging. It would allow more easily half-transparent preview of drag since support for transparent `Window` is varying on different platforms. The gotcha is that preview itself _must_ work outside winddow.

**Comment by BreyerW** [comment]
`InputManager` might be what i need but in Avalonia 11 it is internal so i cant check

**Reply by timunie** [reply]
if you are evil 👿 you could use reflection to access it... BUT be aware, it may break during updates without further notice ^^

**Reply by BreyerW** [reply]
@timunie if push comes to shove i will do that but first wanted to see if theres official solution

**Comment by maxkatz6** [comment]
Pointer and Drag events are routed, and you can add global handlers to them. No need for InputManager.

InputElement.PointerMovedEvent.AddClassHandler ...

**Reply by BreyerW** [reply]
Including during window dragging though? Cause im seeing different in my project 

(one started with `window.BeginDragMove()`)

**Comment by BreyerW** [comment, marked as answer]
Found solution: Ditched `BeginDragMove()` and instead manipulate window position directly but also need to mouse capture something on pointer press (without capture and without dropping `BeginDragMove()` globally listening wont work and i didnt test this particular combo previously - if i remember correctly avalonia 0.10.x had pretty bad lag when moving mouse very fast and directly changing window position so thats why but now with avalonia 11 i dont see issues with window lagging behind mouse even when i try to move as fast as i can with mouse dragging).

EDIT:

Actually there is still some window lag but not as bad as it was.
